src/cogent3/app/composable.py

```diff
@@ -754,3 +757,435 @@ def maker(*args, **kwargs):
+        k: p.default
+        for k, p in init_sig.parameters.items()
+        if k != "self" and (p.default is not no_default)
+    }
+
+    init_vals = {}
+    if arg_order:  ### this line is new, for empty args
+        for i, v in enumerate(args):
+            k, v = arg_order[i], v
+            init_vals[k] = v
+
+    init_vals |= kw_args | kwargs
+    obj._serialisable.kwargs = init_vals
+    return obj
+
+
+def _call(self, val, *args, **kwargs):
+    # logic for trapping call to main
+
+    if val is None:
+        ## new message in place of traceback
+        return NotCompleted("ERROR", self, "unexpected None value as input", source=val)
+
+    if not val:
+        return val
+
+    if self.app_type is not LOADER and self.input:  # passing to connected app
+        val = self.input(val, *args, **kwargs)
+        if not val:  ### moved this if into above if
+            return val
+
+    type_checked = self._validate_data_type(val)
+    if not type_checked:
+        return type_checked
+
+    try:
+        result = self.main(val, *args, **kwargs)
+    except Exception:
+        result = NotCompleted("ERROR", self, traceback.format_exc(), source=val)
+
+    if not result and not isinstance(result, NotCompleted):
+        msg = (
+            f"The value {result!r} equates to False. "
+            "If unexpected, please post this error message along"
+            f" with the code and data {val!r} as an Issue on the"
+            " github project page."
+        )
+        result = NotCompleted("BUG", f"{self.__class__.__name__}", msg, source=val)
+
+    return result
+
+
+def _getstate(self):
+    data = self._serialisable()
+    return deepcopy(data)
+
+
+def _setstate(self, data):
+    return _new(self.__class__, **data)
+
+
+def _validate_data_type(self, data):
+    """checks data class name matches defined compatible types"""
+    # if not self._data_types:
+    #     # not defined
+    #     return True
+    class_name = data.__class__.__name__
+    data_types = get_constraint_names(next(iter(self._data_types)))
+    valid = class_name in data_types
+    if not valid:
+        msg = f"invalid data type, '{class_name}' not in {', '.join([p.__name__ for p in self._data_types])}"
+        valid = NotCompleted("ERROR", self, message=msg, source=data)
+    return valid
+
+
+class connected:
+    def __init__(self):
+        self.storage = {}
+
+    def __get__(self, instance, owner):
+        if instance.app_type is LOADER:
+            raise TypeError("Looking for input value for Loader app is not permitted.")
+        return self.storage.get(id(instance), None)
+
+    def __set__(self, instance, value):
+        if instance.app_type is WRITER:
+            raise TypeError("Assigning input value for Writer app is not permitted.")
+        self.storage[id(instance)] = value
+
+
+## New variable (app_type)
+def composable(klass=None, *, app_type: AppType = GENERIC):
+    # this validates as known value, raises ValueError otherwise
```

Comment:
delete comment

We need to add docstrings to all these classes just before we merge into the `develop` branch